matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/rendezvous/Rendezvous.kt

```diff
@@ -0,0 +1,225 @@
+        return authenticationService.loginUsingQrLoginToken(hsConfig, token)
+    }
+
+    @Throws(RendezvousError::class)
+    suspend fun completeVerificationOnNewDevice(session: Session) {
+        val userId = session.myUserId
+        val crypto = session.cryptoService()
+        val deviceId = crypto.getMyDevice().deviceId
+        val deviceKey = crypto.getMyDevice().fingerprint()
+        send(Payload(PayloadType.PROGRESS, outcome = Outcome.SUCCESS, deviceId = deviceId, deviceKey = deviceKey))
+
+        // await confirmation of verification
+        val verificationResponse = receive()
+        if (verificationResponse?.outcome == Outcome.VERIFIED) {
+            val verifyingDeviceId = verificationResponse.verifyingDeviceId
+                    ?: throw RendezvousError("No verifying device id returned", RendezvousFailureReason.ProtocolError)
+            val verifyingDeviceFromServer = crypto.getCryptoDeviceInfo(userId, verifyingDeviceId)
+            if (verifyingDeviceFromServer?.fingerprint() != verificationResponse.verifyingDeviceKey) {
+                Timber.tag(TAG).w(
+                        "Verifying device $verifyingDeviceId key doesn't match: ${
+                            verifyingDeviceFromServer?.fingerprint()
+                        } vs ${verificationResponse.verifyingDeviceKey})"
+                )
+                // inform the other side
+                send(Payload(PayloadType.FINISH, outcome = Outcome.E2EE_SECURITY_ERROR))
+                throw RendezvousError("Key from verifying device doesn't match", RendezvousFailureReason.E2EESecurityIssue)
+            }
+
+            verificationResponse.masterKey?.let { masterKeyFromVerifyingDevice ->
+                // check master key againt what the homeserver told us
+                crypto.crossSigningService().getMyCrossSigningKeys()?.masterKey()?.let { localMasterKey ->
+                    if (localMasterKey.unpaddedBase64PublicKey != masterKeyFromVerifyingDevice) {
+                        Timber.tag(TAG).w("Master key from verifying device doesn't match: $masterKeyFromVerifyingDevice vs $localMasterKey")
+                        // inform the other side
+                        send(Payload(PayloadType.FINISH, outcome = Outcome.E2EE_SECURITY_ERROR))
+                        throw RendezvousError("Master key from verifying device doesn't match", RendezvousFailureReason.E2EESecurityIssue)
+                    }
+                    // set other device as verified
+                    Timber.tag(TAG).i("Setting device $verifyingDeviceId as verified")
+                    crypto.setDeviceVerification(DeviceTrustLevel(locallyVerified = true, crossSigningVerified = false), userId, verifyingDeviceId)
+
+                    Timber.tag(TAG).i("Setting master key as trusted")
+                    crypto.crossSigningService().markMyMasterKeyAsTrusted()
+                } ?: Timber.tag(TAG).w("No local master key so not verifying")
```

Comment:
I think there is a problem here, the other side says what the MSK is and on your side you see nothing. This is a malicious home server droping some of your keys

Reply:
Well spotted. The master key is now always validated even if  not provided by home server.